hc_dpki/src/seed.rs

```diff
@@ -0,0 +1,558 @@
+    pub fn get_seed_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        let mut passphrase = SecBuf::with_insecure_from_string(passphrase);
+        let seed_data: bundle::ReturnBundleData =
+            util::pw_enc(&mut self.seed_buf, &mut passphrase, config)?;
+
+        // convert -> to string -> to base64
+        let seed_data_serialized = json::encode(&seed_data).unwrap();
+        let seed_data_encoded = base64::encode(&seed_data_serialized);
+
+        Ok(bundle::KeyBundle {
+            bundle_type: self.seed_type.clone(),
+            hint,
+            data: seed_data_encoded,
+        })
+    }
+
+    ///  Initialize this seed class with persistence bundle type and private seed
+    ///
+    ///  @param {string} stype - the persistence bundle type
+    ///
+    ///  @param {SecBuf|string} seed - the private seed data (as a buffer or mnemonic)
+    pub fn new(stype: &String, sm: InitializeSeed) -> Self {
+        match sm {
+            SeedInit(s) => Seed {
+                seed_type: stype.clone(),
+                seed_buf: s,
+            },
+            MnemonicInit(phrase) => {
+                let mnemonic = Mnemonic::from_phrase(phrase, Language::English).unwrap();
+                let entropy: &[u8] = mnemonic.entropy();
+                let mut buf = SecBuf::with_insecure(entropy.len());
+                util::convert_array_to_secbuf(entropy, &mut buf);
+                Seed {
+                    seed_type: stype.clone(),
+                    seed_buf: buf,
+                }
+            }
+        }
+    }
+
+    /// Generated a mnemonic for the seed.
+    pub fn get_mnemonic(&mut self) -> Result<String, HolochainError> {
+        let entropy = self.seed_buf.read_lock();
+        let e = &*entropy;
+        let mnemonic = Mnemonic::from_entropy(e, Language::English).unwrap();
+        Ok(mnemonic.phrase().to_string())
+    }
+}
+#[derive(Debug)]
+pub struct DevicePinSeed {
+    s: Seed,
+}
+
+impl DevicePinSeed {
+    /// creates bundle for for the seed
+    pub fn get_bundle(
+        &mut self,
+        passphrase: String,
+        hint: String,
+        config: Option<PwHashConfig>,
+    ) -> Result<bundle::KeyBundle, HolochainError> {
+        Ok(self.s.get_seed_bundle(passphrase, hint, config)?)
+    }
+
+    /// delegate to base struct
+    pub fn new(s: SecBuf) -> Self {
+        DevicePinSeed {
+            s: Seed::new(&"hcDevicePinSeed".to_string(), SeedInit(s)),
+        }
+    }
+
+    /// generate an application keypair given an index based on this seed
+    /// @param {number} index
+    /// @return {Keypair}
+    pub fn get_application_keypair(&mut self, index: u64) -> Result<Keypair, HolochainError> {
+        (index >= 1).ok_or(HolochainError::ErrorGeneric("Invalid index".to_string()))?;
+
+        let mut out_seed = SecBuf::with_insecure(32);
```

Comment:
... i'm less certain about this one. At the very least, this should be made `with_secure`.

the seed isn't exported, and Keypair::new_from_seed below always returns secure secrets... so maybe we can get away without letting the implementor decide here?